### DISCUSSION TITLE
Terminals with complex data type

### DISCUSSION THREAD

**Discussion by Lotes** [eclipse-langium/langium#758] [Ideas]
For Langium terminals it is possible to be defined as RegExp literal.
One feature of regular expressions are [named groups](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Groups_and_Backreferences#using_named_groups). 

The idea is to map named groups to properties. Here is one example for RGB values:
```antlr
//before, parse later
ColorLiteral returns string: /#[a-f0-9]{6}/i;

//after
interface Color {
  red: string;
  green: string;
  blue: string;
}
ColorLiteral returns Color: /#(?<red>[a-f0-9]{2})(?<green>[a-f0-9]{2})(?<blue>[a-f0-9]{2})/i;
```

But maybe it is too much complexity for so little value. But now it is out of my head, more space for other stuff teehee

Values, I see, are:

* less duplicated code on TS side (the regex is duplicated if you need to parse the token again)
* complex types for terminals

**Comment by spoenemann** [comment]
Nice idea! Currently you'd have to do this with a parser rule:

```
ColorLiteral: '#' red=ColorToken green=ColorToken blue=ColorToken;

terminal ColorToken: /#[a-f0-9]{2}/;
```

But this would see the object as four distinct tokens and thus allow whitespace between them. The whitespace could be suppressed with a validation rule that looks at the CST to find "hidden tokens".
